libraries/AP_HAL_PX4/UARTDriver.cpp

```diff
@@ -274,7 +275,10 @@ size_t PX4UARTDriver::write(uint8_t c)
         }
         hal.scheduler->delay(1);
     }
-    return _writebuf.write(&c, 1);
+	_write_sem.take(10);
```

Comment:
Also have to check if you actually got the semaphore before then trying to use the protected item, and before releasing it.